pbft/pbft_service.cpp

```diff
@@ -16,18 +16,28 @@
 
 using namespace bzn;
 
+pbft_service::pbft_service(std::shared_ptr<pbft_failure_detector_base> failure_detector)
+        : failure_detector(std::move(failure_detector))
+{
+}
+
 void
-pbft_service::commit_request(uint64_t sequence, const pbft_request& request)
+pbft_service::commit_operation(std::shared_ptr<pbft_operation> operation)
 {
-    this->waiting_requests[sequence] = request;
+    uint64_t sequence = operation->sequence;
+    this->waiting_requests[sequence] = std::move(operation);
 
     while (this->waiting_requests.count(this->next_request_sequence) > 0)
     {
-        const pbft_request& req = waiting_requests[sequence];
+        std::shared_ptr<pbft_operation>& op = waiting_requests[this->next_request_sequence];
```

Comment:
This could be dangerous if the original shared_ptr becomes invalid?
I would remove the reference.